### DISCUSSION TITLE
TraceQL - How to get average duration of traces ?

### DISCUSSION THREAD

**Discussion by hugoallaine** [grafana/tempo#4354] [Q&A] [answered]
Probably a simple question, but I cannot find this information.
I read the doc and find some way to get traces by avg duration / trace duration but how can I get the avg trace duration from traces ?

In Tempo, I have some traces (2 for example here)
![image](https://github.com/user-attachments/assets/ba7530e5-65f3-4e60-8c96-04fcccad6ebc)

In Grafana, I can see the trace duration from Tempo
![image](https://github.com/user-attachments/assets/015cdb92-d311-453e-90e4-cb225d3fd983)

The goal is to get a the average trace duration (here (2.32+0.941)/2 = 1.6305)

In other words, I'd like to do something like this :
`{ name="my_root_span_that_is_in_all_traces" } | avg(traceDuration)`

I've probably missed this information, thanks in advance for your help

**Comment by joe-elliott** [comment, marked as answer]
`avg_over_time` was [added recently](https://github.com/grafana/tempo/pull/4073) and will be available in Tempo 2.7 in a few months.

The query would look something like:

```
{ trace:rootName="root span name" } | avg_over_time(trace:duration)
```

This should give you what you are looking for. In Tempo 2.6 the best way to get this information would be to use the [metrics generators](https://grafana.com/docs/tempo/latest/metrics-generator/)

**Reply by hugoallaine** [reply]
Thank you for your answer !
I tried metrics generators to do that by summing the average time of all root span (my traces don't have a root span that takes all the time of the trace) but it's not accurate in my setup because I have async mq (Pulsar) in the middle of the whole process and I cannot have the waiting time between two different "root" span with TraceQL. (or I missed another thing?)

I'll use your solution when Tempo 2.7 will be available.

**Comment by mdisibio** [comment]
It's a little different than average but you can get the p50 already with:

```
{ name="my_root_span_that_is_in_all_traces" } | quantile_over_time(duration, 0.5)
```

**Reply by joe-elliott** [reply]
good point! didn't think about that. definitely easier than setting up metrics generator

**Reply by hugoallaine** [reply]
> It's a little different than average but you can get the p50 already with:
> 
> ```
> { name="my_root_span_that_is_in_all_traces" } | quantile_over_time(duration, 0.5)
> ```

Thank you too for your answer !
This is a useful metric, I'll use it in my setup to complement the average.
